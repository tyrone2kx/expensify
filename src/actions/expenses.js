import uuid from 'uuid';
import database from '../firebase/firebase';



// ADD_EXPENSE ACTION GENERATOR

export const addExpense = (expense) => ({
    type: 'ADD_EXPENSE',
    expense
});



// START ADD EXPENSE

export const startAddExpense = (expenseData = {}) => {
    return (dispatch) => {
        const {
            description = '', 
            note = '', 
            amount = 0, 
            createdAt = 0
        } = expenseData; // this is same as setting the defaults inside the arguement brackets

        const expense = { description, note, amount, createdAt };

        return database.ref('expenses').push(expense).then((ref) => {
            dispatch(addExpense({
                id: ref.key,
                ...expense
            }));
        });
    }
}






// REMOVE_EXPENSE ACTION GENERATOR

export const removeExpense = ({ id } = {}) => ({
    type: 'REMOVE_EXPENSE',
    id // id: id
    
});


// EDIT_EXPENSE

export const editExpense = (id, updates) => ({
    type: 'EDIT_EXPENSE',
    id,
    updates
});
