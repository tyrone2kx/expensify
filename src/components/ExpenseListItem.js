import React from 'react';
import { Link } from 'react-router-dom';
import moment from 'moment';
import numeral from 'numeral';



const ExpenseListItem = ({ id, description, amount, createdAt}) => (
    <div>
    <Link to={`/edit/${id}`}>
        <h3>{description}</h3>
        <p>
        {numeral(amount / 100).format('$0,0.00')} - 
        
        {moment(createdAt).format('MMMM Do, YYYY')}
        
        </p>
    </Link>

        
    </div>
)


export default ExpenseListItem;