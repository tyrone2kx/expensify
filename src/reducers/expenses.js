// EXPENSE REDUCER

const expensesReducerDefaultState = [];

const expenseReducer = (state = expensesReducerDefaultState, action) => {
    switch (action.type) {
        case 'ADD_EXPENSE':
            // return state.concat(action.expense) [this can be achieved using the es6 spread operator asshown below. It doesn't change the original state.]
            return [...state, action.expense]
        
        case 'REMOVE_EXPENSE':
            return state.filter(({ id }) => id !== action.id);

        case 'EDIT_EXPENSE':
            return state.map((expense) => {
                if (expense.id === action.id) {
                    return {
                        ...expense,
                        ...action.updates
                    };
                } else {
                    return expense; // as good as saying "do nothing!"
                }
            });

    default:
        return state;
    }

}



export default expenseReducer;