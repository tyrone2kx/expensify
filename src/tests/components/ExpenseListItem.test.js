import React from 'react';
import { shallow } from 'enzyme';
import ExpenseListItem from '../../components/ExpenseListItem';
import expenses from '../fixtures/expenses';



test('should return expense list with expense', () => {
    const wrapper = shallow(<ExpenseListItem key={expenses[0].id} {...expenses[0]} />);
    expect(wrapper).toMatchSnapshot();
});