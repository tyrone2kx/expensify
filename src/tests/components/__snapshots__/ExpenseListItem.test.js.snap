// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should return expense list with expense 1`] = `
<div>
  <Link
    replace={false}
    to="/edit/1"
  >
    <h3>
      gum
    </h3>
    <p>
      195
       - 
      0
    </p>
  </Link>
</div>
`;
